service-base/src/main/java/org/eclipse/hono/service/HonoBeanNameGenerator.java

```diff
@@ -0,0 +1,28 @@
+/**
+ * Copyright (c) 2017 Bosch Software Innovations GmbH.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Bosch Software Innovations GmbH - initial creation
+ */
+
+package org.eclipse.hono.service;
+
+import org.springframework.beans.factory.config.BeanDefinition;
+import org.springframework.beans.factory.support.BeanDefinitionRegistry;
+import org.springframework.beans.factory.support.BeanNameGenerator;
+
+/**
+ * Generates bean names that are identical to the class name.
+ */
+public class HonoBeanNameGenerator implements BeanNameGenerator {
```

Comment:
I changed my mind. Option 3 is not the Spring way. Bean names and class names are not necessarily connected, it is just a default and there are slightly different mechanisms how Spring derives them. (You could have multiple beans of the same type and to make it more complex multiple names for one bean).
Spring needs us to name beans with a constant value. The cleanest solution would be to explicitly name the bean as well as the target bean for the `ObjectFactory.`

My proposal is to use constants for the bean names here and combine it with Java-based configuration (which is IMHO preferable anyway) instead of of annotation-based configuration. This would make it short and readable. Example: 
Remove `@Component` from `DeviceRegistryAmqpServer` and use this configuration pattern:
```
@Configuration
public class ApplicationConfig {

    private static final String BEAN_NAME_DEVICE_REGISTRY_AMQP_SERVER = "deviceRegistryAmqpServer"; // Just a name, could be FooBar.

    @Qualifier("amqp")
    @Bean
    public ObjectFactoryCreatingFactoryBean amqpServerFactory() {
        ObjectFactoryCreatingFactoryBean factory = new ObjectFactoryCreatingFactoryBean();
        factory.setTargetBeanName(BEAN_NAME_DEVICE_REGISTRY_AMQP_SERVER);
        return factory;
    }

    @Bean(BEAN_NAME_DEVICE_REGISTRY_AMQP_SERVER)
    @Scope("prototype")
    public DeviceRegistryAmqpServer deviceRegistryAmqpServer() {
        return new DeviceRegistryAmqpServer();
    }
...
}
```